ksql-engine/src/main/java/io/confluent/ksql/query/id/HybridQueryIdGenerator.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.query.id;
+
+import com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Contains both a Sequential Query Id Generator and a Specific QueryId Generator
+ * Can toggle between the two depending on which one is currently activated
+ * The query id generation method was changed from using
+ */
+public class HybridQueryIdGenerator implements QueryIdGenerator {
+
+  private final SequentialQueryIdGenerator legacyGenerator;
+  private final SpecificQueryIdGenerator newGenerator;
+  private QueryIdGenerator activeGenerator;
+
+  public HybridQueryIdGenerator() {
+    this(0L);
+  }
+
+  public HybridQueryIdGenerator(final long initialValue) {
+    this.legacyGenerator = new SequentialQueryIdGenerator(initialValue);
+    this.newGenerator = new SpecificQueryIdGenerator();
+    this.activeGenerator = this.legacyGenerator;
+  }
+
+  @VisibleForTesting
+  HybridQueryIdGenerator(
+      final SequentialQueryIdGenerator sequentialQueryIdGenerator,
+      final SpecificQueryIdGenerator specificQueryIdGenerator
+  ) {
+    this.legacyGenerator = sequentialQueryIdGenerator;
+    this.newGenerator = specificQueryIdGenerator;
+    this.activeGenerator = this.legacyGenerator;
+  }
+
+  public void activateNewGenerator(final long nextId) {
+    newGenerator.setNextId(nextId);
+    this.activeGenerator = newGenerator;
+  }
+
+  public void activateLegacyGenerator() {
```

Comment:
I don't think we should allow this method (e.g. as soon as we start using the new one, we should fail if we ever activate the legacy one). Otherwise we run a larger risk of generating duplicate queryIDs

Reply:
Andy mentioned we should have something to potentially switch back in this comment.
https://github.com/confluentinc/ksql/pull/3354#discussion_r326595453

Reply:
I disagree with Andy, but not strongly enough to block this so it's your choice

Reply:
As long as we can ensure that for any sequence of commands, the query id is always increasing, there shouldn't be any duplicate queryIDs. When activating the legacy, we could call legacyGenerator.setCounter() or setNextID() (TODO) in order to "jump" ahead to whatever offset/id the newGenerator was at + 1 and start the counter from there.

Reply:
There's nothing perfect we can do here. If we throw an error in this case, we may be skipping statements that were executed by <=5.3 KSQLs. There's also the issue of forward compatibility - commands that we write with 5.4 will get different ids if run with 5.3.

I think we should just call out in the docs that this upgrade must be disruptive - you have to shut down all your KSQLs, upgrade, and then restart them.

Reply:
@rodesai should that note be made here? 
https://github.com/confluentinc/ksql/blob/master/docs/installation/upgrading.rst
There's no section yet about upgrading to 5.4, so should I wait until there is one to add it? I think the upgrade doesn't have to be completely disruptive also, we can warn that users shouldn't create new Push Queries (CSAS/CTAS) during the process of upgrading their servers.

Reply:
https://github.com/confluentinc/ksql/pull/3559